00_Base/src/repository/TokensRepository.ts

```diff
@@ -245,50 +168,110 @@ export class TokensRepository extends SequelizeRepository<OcpiToken> {
     authorizationId: number,
     tokenDto: TokenDTO,
   ): Promise<OcpiToken> {
+    const baseTokenData = {
+      visual_number: tokenDto.visual_number,
+      issuer: tokenDto.issuer,
+      whitelist: tokenDto.whitelist,
+      default_profile_type: tokenDto.default_profile_type,
+      energy_contract: tokenDto.energy_contract,
+      last_updated: tokenDto.last_updated,
+      type: tokenDto.type,
+    }
+
     const [savedOcpiToken, ocpiTokenCreated] = await this._readOrCreateByQuery({
       where: {
         authorization_id: authorizationId,
       },
       defaults: {
-        party_id: tokenDto.party_id,
-        country_code: tokenDto.country_code,
-        visual_number: tokenDto.visual_number,
-        issuer: tokenDto.issuer,
-        whitelist: tokenDto.whitelist,
-        default_profile_type: tokenDto.default_profile_type,
-        energy_contract: tokenDto.energy_contract,
-        last_updated: tokenDto.last_updated,
+        ...baseTokenData,
         authorization_id: authorizationId,
-        type: tokenDto.type,
+        country_code: tokenDto.country_code,
+        party_id: tokenDto.party_id,
       },
     });
+
     // TODO confirm all updatable properties
     if (!ocpiTokenCreated) {
-      await this._updateByKey(
-        {
-          visual_number: tokenDto.visual_number,
-          issuer: tokenDto.issuer,
-          whitelist: tokenDto.whitelist,
-          default_profile_type: tokenDto.default_profile_type,
-          energy_contract: tokenDto.energy_contract,
-          last_updated: tokenDto.last_updated,
-          type: tokenDto.type,
-        },
-        savedOcpiToken.id,
-      );
+      await this._updateByKey(baseTokenData, savedOcpiToken.id);
     }
 
     return savedOcpiToken;
   }
 
-  getMatchingAuth(
+  private getMatchingAuth(
     id: number,
     ocppAuths: Authorization[],
   ): Authorization | undefined {
     return ocppAuths.find((ocppAuth) => ocppAuth.id === id);
   }
+
+  private async getOcppAuths(partialToken: TokenDTO | SingleTokenRequest): Promise<Authorization[]> {
+    return await this.authorizationRepository.readAllByQuerystring({
+      idToken: partialToken.uid!,
+      type: OcpiTokensMapper.mapOcpiTokenTypeToOcppIdTokenType(partialToken.type),
+    });
+  }
+
+  private async getOcpiTokenFromAuths(queryParams: SingleTokenRequest | TokenDTO, ocppAuths: Authorization[]): Promise<OcpiToken | undefined> {
+    return await this.readOnlyOneByQuery({
+      where: {
+        authorization_id: {
+          [Op.in]: ocppAuths.map((ocppAuth) => ocppAuth.id),
+        },
+        country_code: queryParams.country_code,
+        party_id: queryParams.party_id,
+        type: queryParams.type,
+      },
+    });
+  }
+
+  private async updateOcpiToken(existingOcpiToken: OcpiToken, partialToken: Partial<OcpiToken>, transaction: SequelizeTransaction): Promise<OcpiToken> {
+    const [updatedCount, [updatedToken]] = await OcpiToken.update(partialToken, {
```

Comment:
should this be `const [updatedCount, updatedToken]` instead so that `updatedToken` is not in a list, or is the list intentional?

Reply:
I meant to always take the first token in the update but good point, I think this can NPE right? Will change!